src/include/storage/write_ahead_log/log_manager.h

```diff
@@ -68,25 +109,97 @@ class LogManager {
-  // optimization we applied to the GC queue.
+  //  optimization we applied to the GC queue.
   std::queue<RecordBufferSegment *> flush_queue_;
 
   // These do not need to be thread safe since the only thread adding or removing from it is the flushing thread
   std::vector<std::pair<transaction::callback_fn, void *>> commits_in_buffer_;
 
+  // System path for log file
+  const char *log_file_path_;
+  // This stores all the buffers the serializer or the log writer threads use
+  std::vector<BufferedLogWriter> buffers_;
+  // This is the buffer the serializer thread will write to
+  BufferedLogWriter *buffer_to_write_;
+  // The queue containing empty buffers which the serializer thread will use
+  common::ConcurrentBlockingQueue<BufferedLogWriter *> empty_buffer_queue_;
+  // The queue containing filled buffers pending flush to the disk
+  common::ConcurrentBlockingQueue<BufferedLogWriter *> filled_buffer_queue_;
+
+  // The log writer object which flushes filled buffers to the disk
+  LogWriter *log_writer_;
+  // Flag used by the serializer thread to signal shutdown to the log writer thread
+  volatile bool run_log_writer_thread_;
+  // Flag used by the serializer thread to signal the log writer thread to persist the data on disk
+  volatile bool do_persist_;
+
+  // Synchronisation primitives to synchronise persisting buffers to disk
+  std::mutex persist_lock_;
+  std::condition_variable persist_cv_;
+  // Condition variable to signal writer thread to wake up and flush buffers to disk or if shutdown has initiated,
+  // then quit
+  std::condition_variable wake_writer_thread_cv_;
+
+  /**
+   * Serialize out the record to the log
+   * @param record the redo record to serialise
+   */
   void SerializeRecord(const LogRecord &record);
 
+  /**
+   * Serialize out the task buffer to the log
+   * @param task_buffer the iterator to the redo buffer to be serialized
+   */
+  void SerializeTaskBuffer(IterableBufferSegment<LogRecord> *task_buffer);
+
+  /**
+   * Used by the serializer thread to get a buffer to serialize data to
+   * @return buffer to write to
+   */
+  BufferedLogWriter *GetBufferToWrite() {
+    if (buffer_to_write_ == nullptr) {
+      empty_buffer_queue_.Dequeue(&buffer_to_write_);
+    }
+    return buffer_to_write_;
+  }
+
+  /**
+   * Serialize the data pointed to by val to a buffer
+   * @tparam T Type of the value
+   * @param val The value to write to the buffer
+   */
   template <class T>
   void WriteValue(const T &val) {
-    out_.BufferWrite(&val, sizeof(T));
+    WriteValue(&val, sizeof(T));
+  }
+
+  /**
+   * Serialize the data pointed to by val to a buffer
+   * @param val the value
+   * @param size size of the value to serialize
+   */
+  void WriteValue(const void *val, uint32_t size);
+
+  /**
+   * Mark the current buffer that the serializer thread is writing to as filled
+   */
+  void MarkBufferFull() {
```

Comment:
You are not merely marking the buffer as full, but also doing stuff with it, right? The name should probably reflect that. (e.g. `HandFilledBufferToWriter`?)